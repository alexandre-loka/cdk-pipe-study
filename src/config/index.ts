export { Config } from "./types";
export { config as mainConfig } from "./config.main";
export { config as devConfig } from "./config.dev";
