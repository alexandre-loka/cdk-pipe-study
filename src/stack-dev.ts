import { Stack, StackProps, Tags } from "aws-cdk-lib";
import { Construct } from "constructs";
import * as codecommit from "aws-cdk-lib/aws-codecommit";

export class InfraDev extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    new codecommit.Repository(this, "AppRepo", { repositoryName: "app.coelhor.dev" });
    new codecommit.Repository(this, "LambdasRepo", { repositoryName: "lambdas" });

    Tags.of(this).add("Project", "Study project");
    Tags.of(this).add("Team", "Alexandre Ramos");
  }
}
