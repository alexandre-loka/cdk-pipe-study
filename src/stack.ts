import { Stack, StackProps, Tags } from "aws-cdk-lib";
import { Construct } from "constructs";

export class CHANGE_ME_Stack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // Please configure the proper tags for this project
    Tags.of(this).add("Project", "Name of project");
    Tags.of(this).add("Team", "Name of team");
  }
}
