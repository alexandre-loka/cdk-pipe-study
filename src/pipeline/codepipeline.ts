import { Stack, StackProps, Stage, StageProps, Tags } from "aws-cdk-lib";
import { CodePipeline, CodePipelineSource, ShellStep } from "aws-cdk-lib/pipelines";
import { Construct } from "constructs";
import { InfraBase } from "../stack";

export class PipelineStage extends Stage {
  constructor(scope: Construct, id: string, props?: StageProps) {
    super(scope, id, props);

    new InfraBase(this, "StageStack", {});
  }
}

/**
 * Stack to that creates a CDK CodePipeline pipeline
 */
export class CodePipelineStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // Pipeline source: branch and Github org/repository
    const branch = "main";
    const githubRepository = "org/repo";
    /**
     * A Codestar connection to Github must be manually created using the AWS console
     * see https://docs.aws.amazon.com/codepipeline/latest/userguide/update-github-action-connections.html
     */
    const githubConnection = CodePipelineSource.connection(githubRepository, branch, {
      connectionArn:
        "arn:aws:codestar-connections:us-east-1:222222222222:connection/7d2469ff-514a-4e4f-9003-5ca4a43cdc41",
    });

    const pipeline = new CodePipeline(this, "PipelineStack", {
      pipelineName: "example-codepipeline",
      synth: new ShellStep("Synth", {
        input: githubConnection,
        commands: ["yarn", "yarn run cdk synth"],
        primaryOutputDirectory: "cdk.out",
      }),
      selfMutation: true, // Whether the pipeline will update itself.
      crossAccountKeys: false, // Create KMS keys for the artifact buckets, allowing cross-account deployments.
    });

    // A pipeline can have multiple stages (for ex. dev, staging, prod)
    pipeline.addStage(new PipelineStage(this, "PipelineStage", {}));

    // Please configure the proper tags for this project
    Tags.of(this).add("Project", "Name of project");
    Tags.of(this).add("Team", "Name of team");
  }
}
