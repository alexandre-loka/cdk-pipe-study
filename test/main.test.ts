import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";

import { CHANGE_ME_Stack } from "../src/stack";
import { CodePipelineStack } from "../src/pipeline/codepipeline";

// Test for single stack
test("snapshot for CHANGE_ME_Stack matches previous state", () => {
  const app = new cdk.App();
  const stack = new CHANGE_ME_Stack(app, "MyTestStack");

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});

// Test for pipeline stack
test("snapshot for CodePipelineStack matches previous state", () => {
  const app = new cdk.App();
  const stack = new CodePipelineStack(app, "MyCodePipelineStack");

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
