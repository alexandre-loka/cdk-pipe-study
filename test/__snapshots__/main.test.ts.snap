// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot for CHANGE_ME_Stack matches previous state 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`snapshot for CodePipelineStack matches previous state 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "PipelineStackCodeBuildActionRoleBF300C35": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "PipelineStackPipelineRoleC1A5441C",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": [
          {
            "Key": "Project",
            "Value": "Name of project",
          },
          {
            "Key": "Team",
            "Value": "Name of team",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineStackCodeBuildActionRoleDefaultPolicy4F234311": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "PipelineStackPipelineBuildSynthCdkBuildProject699C1502",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "PipelineStackUpdatePipelineSelfMutation0E4771C6",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineStackCodeBuildActionRoleDefaultPolicy4F234311",
        "Roles": [
          {
            "Ref": "PipelineStackCodeBuildActionRoleBF300C35",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineStackPipelineArtifactsBucket02A4617B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "Project",
            "Value": "Name of project",
          },
          {
            "Key": "Team",
            "Value": "Name of team",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "PipelineStackPipelineArtifactsBucketPolicy55C10FE4": {
      "Properties": {
        "Bucket": {
          "Ref": "PipelineStackPipelineArtifactsBucket02A4617B",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineStackPipelineArtifactsBucket02A4617B",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineStackPipelineArtifactsBucket02A4617B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":role/cdk-hnb659fds-deploy-role-",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      "-",
                      {
                        "Ref": "AWS::Region",
                      },
                    ],
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineStackPipelineArtifactsBucket02A4617B",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineStackPipelineArtifactsBucket02A4617B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "PipelineStackPipelineBuildSynthCdkBuildProject699C1502": {
      "Properties": {
        "Artifacts": {
          "Type": "CODEPIPELINE",
        },
        "Cache": {
          "Type": "NO_CACHE",
        },
        "Description": "Pipeline step MyCodePipelineStack/Pipeline/Build/Synth",
        "EncryptionKey": "alias/aws/s3",
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:7.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": {
          "Fn::GetAtt": [
            "PipelineStackPipelineBuildSynthCdkBuildProjectRole18730EFB",
            "Arn",
          ],
        },
        "Source": {
          "BuildSpec": "{
  "version": "0.2",
  "phases": {
    "build": {
      "commands": [
        "yarn",
        "yarn run cdk synth"
      ]
    }
  },
  "artifacts": {
    "base-directory": "cdk.out",
    "files": "**/*"
  }
}",
          "Type": "CODEPIPELINE",
        },
        "Tags": [
          {
            "Key": "Project",
            "Value": "Name of project",
          },
          {
            "Key": "Team",
            "Value": "Name of team",
          },
        ],
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "PipelineStackPipelineBuildSynthCdkBuildProjectRole18730EFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": [
          {
            "Key": "Project",
            "Value": "Name of project",
          },
          {
            "Key": "Team",
            "Value": "Name of team",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineStackPipelineBuildSynthCdkBuildProjectRoleDefaultPolicyAD975840": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "PipelineStackPipelineBuildSynthCdkBuildProject699C1502",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "PipelineStackPipelineBuildSynthCdkBuildProject699C1502",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    {
                      "Ref": "PipelineStackPipelineBuildSynthCdkBuildProject699C1502",
                    },
                    "-*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineStackPipelineArtifactsBucket02A4617B",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineStackPipelineArtifactsBucket02A4617B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineStackPipelineBuildSynthCdkBuildProjectRoleDefaultPolicyAD975840",
        "Roles": [
          {
            "Ref": "PipelineStackPipelineBuildSynthCdkBuildProjectRole18730EFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineStackPipelineFC132F4C": {
      "DependsOn": [
        "PipelineStackPipelineRoleDefaultPolicyB9603F04",
        "PipelineStackPipelineRoleC1A5441C",
      ],
      "Properties": {
        "ArtifactStore": {
          "Location": {
            "Ref": "PipelineStackPipelineArtifactsBucket02A4617B",
          },
          "Type": "S3",
        },
        "Name": "example-codepipeline",
        "RestartExecutionOnUpdate": true,
        "RoleArn": {
          "Fn::GetAtt": [
            "PipelineStackPipelineRoleC1A5441C",
            "Arn",
          ],
        },
        "Stages": [
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "CodeStarSourceConnection",
                  "Version": "1",
                },
                "Configuration": {
                  "BranchName": "main",
                  "ConnectionArn": "arn:aws:codestar-connections:us-east-1:222222222222:connection/7d2469ff-514a-4e4f-9003-5ca4a43cdc41",
                  "FullRepositoryId": "org/repo",
                },
                "Name": "org_repo",
                "OutputArtifacts": [
                  {
                    "Name": "org_repo_Source",
                  },
                ],
                "RoleArn": {
                  "Fn::GetAtt": [
                    "PipelineStackPipelineSourceorgrepoCodePipelineActionRoleCE6EA6CB",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Source",
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": {
                  "EnvironmentVariables": "[{"name":"_PROJECT_CONFIG_HASH","type":"PLAINTEXT","value":"dda1e15e63498619dad141469baa5a3cd467ba25b21f0887166bfe2881aa8f65"}]",
                  "ProjectName": {
                    "Ref": "PipelineStackPipelineBuildSynthCdkBuildProject699C1502",
                  },
                },
                "InputArtifacts": [
                  {
                    "Name": "org_repo_Source",
                  },
                ],
                "Name": "Synth",
                "OutputArtifacts": [
                  {
                    "Name": "Synth_Output",
                  },
                ],
                "RoleArn": {
                  "Fn::GetAtt": [
                    "PipelineStackCodeBuildActionRoleBF300C35",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Build",
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": {
                  "EnvironmentVariables": "[{"name":"_PROJECT_CONFIG_HASH","type":"PLAINTEXT","value":"53173b1726ea2906c83e60d2fbfc5a806164560c2e6cf07c0067a5e8b269afe9"}]",
                  "ProjectName": {
                    "Ref": "PipelineStackUpdatePipelineSelfMutation0E4771C6",
                  },
                },
                "InputArtifacts": [
                  {
                    "Name": "Synth_Output",
                  },
                ],
                "Name": "SelfMutate",
                "RoleArn": {
                  "Fn::GetAtt": [
                    "PipelineStackCodeBuildActionRoleBF300C35",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "UpdatePipeline",
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Deploy",
                  "Owner": "AWS",
                  "Provider": "CloudFormation",
                  "Version": "1",
                },
                "Configuration": {
                  "ActionMode": "CHANGE_SET_REPLACE",
                  "Capabilities": "CAPABILITY_NAMED_IAM,CAPABILITY_AUTO_EXPAND",
                  "ChangeSetName": "PipelineChange",
                  "RoleArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":iam::",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":role/cdk-hnb659fds-cfn-exec-role-",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        "-",
                        {
                          "Ref": "AWS::Region",
                        },
                      ],
                    ],
                  },
                  "StackName": "PipelineStage-StageStack",
                  "TemplateConfiguration": "Synth_Output::assembly-MyCodePipelineStack-PipelineStage/MyCodePipelineStackPipelineStageStageStackBB26692B.template.json.config.json",
                  "TemplatePath": "Synth_Output::assembly-MyCodePipelineStack-PipelineStage/MyCodePipelineStackPipelineStageStageStackBB26692B.template.json",
                },
                "InputArtifacts": [
                  {
                    "Name": "Synth_Output",
                  },
                ],
                "Name": "Prepare",
                "RoleArn": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":role/cdk-hnb659fds-deploy-role-",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      "-",
                      {
                        "Ref": "AWS::Region",
                      },
                    ],
                  ],
                },
                "RunOrder": 1,
              },
              {
                "ActionTypeId": {
                  "Category": "Deploy",
                  "Owner": "AWS",
                  "Provider": "CloudFormation",
                  "Version": "1",
                },
                "Configuration": {
                  "ActionMode": "CHANGE_SET_EXECUTE",
                  "ChangeSetName": "PipelineChange",
                  "StackName": "PipelineStage-StageStack",
                },
                "Name": "Deploy",
                "RoleArn": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":role/cdk-hnb659fds-deploy-role-",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      "-",
                      {
                        "Ref": "AWS::Region",
                      },
                    ],
                  ],
                },
                "RunOrder": 2,
              },
            ],
            "Name": "PipelineStage",
          },
        ],
        "Tags": [
          {
            "Key": "Project",
            "Value": "Name of project",
          },
          {
            "Key": "Team",
            "Value": "Name of team",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "PipelineStackPipelineRoleC1A5441C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": [
          {
            "Key": "Project",
            "Value": "Name of project",
          },
          {
            "Key": "Team",
            "Value": "Name of team",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineStackPipelineRoleDefaultPolicyB9603F04": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineStackPipelineArtifactsBucket02A4617B",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineStackPipelineArtifactsBucket02A4617B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "PipelineStackPipelineSourceorgrepoCodePipelineActionRoleCE6EA6CB",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "PipelineStackCodeBuildActionRoleBF300C35",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":iam::",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":role/cdk-hnb659fds-deploy-role-",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    "-",
                    {
                      "Ref": "AWS::Region",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineStackPipelineRoleDefaultPolicyB9603F04",
        "Roles": [
          {
            "Ref": "PipelineStackPipelineRoleC1A5441C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineStackPipelineSourceorgrepoCodePipelineActionRoleCE6EA6CB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": [
          {
            "Key": "Project",
            "Value": "Name of project",
          },
          {
            "Key": "Team",
            "Value": "Name of team",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineStackPipelineSourceorgrepoCodePipelineActionRoleDefaultPolicyADFF0A26": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codestar-connections:UseConnection",
              "Effect": "Allow",
              "Resource": "arn:aws:codestar-connections:us-east-1:222222222222:connection/7d2469ff-514a-4e4f-9003-5ca4a43cdc41",
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineStackPipelineArtifactsBucket02A4617B",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineStackPipelineArtifactsBucket02A4617B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:PutObjectAcl",
                "s3:PutObjectVersionAcl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "PipelineStackPipelineArtifactsBucket02A4617B",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineStackPipelineSourceorgrepoCodePipelineActionRoleDefaultPolicyADFF0A26",
        "Roles": [
          {
            "Ref": "PipelineStackPipelineSourceorgrepoCodePipelineActionRoleCE6EA6CB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineStackUpdatePipelineSelfMutation0E4771C6": {
      "Properties": {
        "Artifacts": {
          "Type": "CODEPIPELINE",
        },
        "Cache": {
          "Type": "NO_CACHE",
        },
        "Description": "Pipeline step MyCodePipelineStack/Pipeline/UpdatePipeline/SelfMutate",
        "EncryptionKey": "alias/aws/s3",
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:7.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "Name": "example-codepipeline-selfupdate",
        "ServiceRole": {
          "Fn::GetAtt": [
            "PipelineStackUpdatePipelineSelfMutationRole759DEB9D",
            "Arn",
          ],
        },
        "Source": {
          "BuildSpec": "{
  "version": "0.2",
  "phases": {
    "install": {
      "commands": [
        "npm install -g aws-cdk@2"
      ]
    },
    "build": {
      "commands": [
        "cdk -a . deploy MyCodePipelineStack --require-approval=never --verbose"
      ]
    }
  }
}",
          "Type": "CODEPIPELINE",
        },
        "Tags": [
          {
            "Key": "Project",
            "Value": "Name of project",
          },
          {
            "Key": "Team",
            "Value": "Name of team",
          },
        ],
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "PipelineStackUpdatePipelineSelfMutationRole759DEB9D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": [
          {
            "Key": "Project",
            "Value": "Name of project",
          },
          {
            "Key": "Team",
            "Value": "Name of team",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineStackUpdatePipelineSelfMutationRoleDefaultPolicy00BCE19B": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "PipelineStackUpdatePipelineSelfMutation0E4771C6",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      {
                        "Ref": "PipelineStackUpdatePipelineSelfMutation0E4771C6",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    {
                      "Ref": "PipelineStackUpdatePipelineSelfMutation0E4771C6",
                    },
                    "-*",
                  ],
                ],
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Condition": {
                "ForAnyValue:StringEquals": {
                  "iam:ResourceTag/aws-cdk:bootstrap-role": [
                    "image-publishing",
                    "file-publishing",
                    "deploy",
                  ],
                },
              },
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:*:iam::",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":role/*",
                  ],
                ],
              },
            },
            {
              "Action": "cloudformation:DescribeStacks",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "s3:ListBucket",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineStackPipelineArtifactsBucket02A4617B",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineStackPipelineArtifactsBucket02A4617B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineStackUpdatePipelineSelfMutationRoleDefaultPolicy00BCE19B",
        "Roles": [
          {
            "Ref": "PipelineStackUpdatePipelineSelfMutationRole759DEB9D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
